test/unit/kernel/src/KernelThrow.cpp

```diff
@@ -0,0 +1,85 @@
+ */
+
+#include <alpaka/core/RuntimeMacros.hpp>
+#include <alpaka/test/acc/TestAccs.hpp>
+#include <alpaka/test/queue/Queue.hpp>
+
+#include <catch2/catch_template_test_macros.hpp>
+#include <catch2/catch_test_macros.hpp>
+
+class KernelWithThrow
+{
+public:
+    template<typename TAcc>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc) const -> void
+    {
+        using Idx = alpaka::Idx<TAcc>;
+        using Dim = alpaka::Dim<TAcc>;
+        using Vec = alpaka::Vec<Dim, Idx>;
+        Vec const globalThreadIdx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        Vec const globalThreadExtent = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
+
+        // Map the three dimensional thread index into a one dimensional thread index space.
+        auto const linearizedGlobalThreadIdx = alpaka::mapIdx<1u>(globalThreadIdx, globalThreadExtent);
+
+        if(static_cast<unsigned>(linearizedGlobalThreadIdx[0u]) == 0u)
+        {
+            printf("linearizedGlobalThreadIdx: %u\n", linearizedGlobalThreadIdx[0]);
+            // Throw abort or std::runtime_error depending on acc type
+            ALPAKA_DEVICE_THROW("Exception thrown by the kernel");
+        }
+        alpaka::syncBlockThreads(acc);
+    }
+};
+
+template<typename T, typename Acc>
+void checkThrow()
+{
+    if constexpr(alpaka::accMatchesTags<Acc, T>)
+    {
+        using Idx = alpaka::Idx<Acc>;
+        using Dim = alpaka::Dim<Acc>;
+        using Vec = alpaka::Vec<Dim, Idx>;
+        using QueueProperty = alpaka::Blocking;
+        using Queue = alpaka::Queue<Acc, QueueProperty>;
+
+        auto const platformAcc = alpaka::Platform<Acc>{};
+        auto const devAcc = alpaka::getDevByIdx(platformAcc, 0);
+
+        Queue queue(devAcc);
+        auto const workDiv = alpaka::WorkDivMembers<Dim, Idx>{Vec{8}, Vec{1}, Vec{1}};
+
+        KernelWithThrow kernel;
+        auto runner = [&]() { alpaka::exec<Acc>(queue, workDiv, kernel); };
+        try
+        {
+            runner();
+        }
+        catch(std::runtime_error& e)
+        {
+            CHECK(e.what() == std::string("Exception thrown by the kernel"));
+        }
+        catch(std::exception& e)
+        {
+            FAIL(std::string("Wrong exception type thrown in kernel:") + e.what());
+        }
+    }
+}
+
+using TestAccs = alpaka::test::EnabledAccs<alpaka::DimInt<1u>, std::uint32_t>;
+
+TEMPLATE_LIST_TEST_CASE("ThrowForCpuThreadAndSerial", "[kernel]", TestAccs)
+{
+    using Acc = TestType;
+    // Test runtime-error exceptions. Aborts can not be catched.
+    checkThrow<alpaka::TagCpuThreads, Acc>();
+    checkThrow<alpaka::TagCpuSerial, Acc>();
+
+    // The OpenMP specification mandates that exceptions thrown by some thread must be handled by the same thread and
+    // within the same parallel region. Therefore std::runtime_error aborts the program.
+
+    // checkThrow<alpaka::TagCpuOmp2Blocks, Acc>();
+    // checkThrow<alpaka::TagCpuOmp2Threads, Acc>();
+}
```

Comment:
No because I can catch the "cudaErrorLaunchFailure" string in the exception correctly. But the test eventually makes the other tests fail in the same executable. I will try with putting to a separate exe.

Reply:
Yes, because `__trap()` causes an error that will not be reset.

Maybe one could try destroying the device and re-initialising it, but I'm not sure it's worth it.

Reply:
Yes I don't think it is worth for now. I have moved the test to a standalone file to be able to test cuda backend as well. Thanks.